### PR TITLE
Allow user to create hangouts from slack

### PR DESCRIPTION
**Pull request by gauravchl**
Flow:
1. Slack user post message in slack channel 
2. [DONE]Slack send this message to Codebuddies Meteor server.
3. [DONE]Meteor server read this message using Event subscription in real time https://api.slack.com/events-api 
4. [DONE]Meteor server read the slack user's email and timezone who posted the above message using slack's web API
5. [DONE]Meteor server find the user in it's db whos email is same as the user who posted the message in slack. If user is found then it process to next step.
6. [DONE] Meteor server parse the message and read topic, start/end date and create hangout.
7. [DONE] On successful hangout creation Meteor server send back a message to slack channel.
8. [DONE] For now don't allow one user to create more than 5 hangouts per day from slack.
9. [DONE] Update response message, it should return hangout url like:
    ```
    Hangout created successfully for 25 minutes, starting from Mon Sep. 3rd 2018. 5:00 pm Asia/Kolkota. 
     Go to https://codebuddies.org/hangout/Tu84XrgG8kzTMxsNy to edit or join the hangout.
      ```

Links: 
- Screencast: https://www.youtube.com/watch?v=2u64CBcBYSA
- Issue: https://github.com/codebuddies/codebuddies/issues/937

### CODE REVIEW

File: server/routes.js [hunk 1]
```diff
@@ -1,18 +1,24 @@
-var bodyParser = require('body-parser');
+var bodyParser = require("body-parser");
+import webhooks from "/server/slack/webhooks.js";
```
Comment (sebbel) [inline server/routes.js]: eslint problems? ping @distalx

File: server/routes.js [hunk 2]
```diff
@@ -1,18 +1,24 @@
...
+  res.setHeader("Access-Control-Allow-Origin", "*");
+  res.setHeader("Access-Control-Allow-Methods", "PUT");
+  res.setHeader("Access-Control-Allow-Headers", "Content-Type");
+  //add in some checking ot make sure re.body.hangoutUrl is a hangout URL
+  Hangouts.update(params._id, { $set: { url: req.body.hangoutUrl } }, function(
+    err,
+    result
+  ) {
+    if (err) {
+      res.statusCode = 404;
```
Comment (sebbel) [inline server/routes.js]: 404 is reserved for page not found. not sure, but I think for a server side error, we should use 500

File: server/slack/webhooks.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+const SLACK_USER_URL = "https://slack.com/api/users.info";
+
+const webhooks = {
+  handleNewEvent(params, req, res) {
+    console.log("webhooks.handleNewEvent", req.headers);
+    const { token, type, challenge, event } = req.body || {};
+    const { slackAppToken } = Meteor.settings;
+
+    if (!token || token !== slackAppToken) {
+      res.statusCode = 404;
```
Comment (sebbel) [inline server/slack/webhooks.js]: see above

File: server/slack/webhooks.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      return;
+    }
+
+    if (type === "event_callback" && event) {
+      res.statusCode = 200;
+      res.end();
+      webhooks.processEvent(event);
+      return;
+    }
+    res.statusCode = 404;
```
Comment (sebbel) [inline server/slack/webhooks.js]: see above

File: server/slack/webhooks.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    res.end();
+  },
+
+  // First get User and it's email and it's timezone from slack API
+  // Find Meteor user whos email is same as slack's one
+  // If Meteor user's and slack user's have same email then create hangout
+  processEvent(event) {
+    console.log("webhook.processEvent", event);
+    const { user: slackUserId, text, type, channel, ts } = event || {};
+    if (type !== "message") return;
```
Comment (sebbel) [inline server/slack/webhooks.js]: eslint should throw here ping @distalx

File: server/routes.js (hunk 1 above)
Comment (gauravchl) [reply server/routes.js]: @sebbel  It's coming from my [Prettier](https://prettier.io/) config. Will revert and disable for this project.

File: server/routes.js (hunk 2 above)
Comment (gauravchl) [reply server/routes.js]: @sebbel +1

File: server/slack/webhooks.js (hunk 5 above)
Comment (gauravchl) [reply server/slack/webhooks.js]: @sebbel Do we use any style guide for codebuddies? I can't find any eslint config.

**Comment by gauravchl** [comment]
@sebbel Thanks for the review!

File: server/slack/webhooks.js (hunk 5 above)
Comment (lpatmo) [reply server/slack/webhooks.js]: @sebbel What do you mean by eslint throwing here? Throwing an error?

@gauravchl We have https://docs.codebuddies.org/code-contribution/coding_style, and Prettier does automatically updates the code in the PR to be a certain style. Let me know if you need help with any of the remaining tasks!

Comment (gauravchl) [reply server/slack/webhooks.js]: @lpatmo Thanks Linda! I will tackle remaining tasks this weekend, i might need some help with setting up slack app on production env.

Comment (lpatmo) [reply server/slack/webhooks.js]: @gauravchl Thanks, sounds great! I will be working on CB too this weekend. Re: hosting -- were you thinking something like a DigitalOcean droplet? We should have funds still in our DigitalOcean account from the DO sponsorship. That's where we're hosting https://github.com/codebuddies/greetbot at the moment. (Not sure if it's possible to build _on top_ of greetbot for the purposes of this integration.)

Comment (lpatmo) [reply server/slack/webhooks.js]: Ah, that makes sense. I can def. add keys to both settings-development.json (so we can test it out) and then to settings-production.json for the release.

**Comment by gauravchl** [comment]
@lpatmo @sebbel 
The bot(cb_jarvis) is ready: https://youtu.be/eDrZC0uv8hA

### How to test on development machine 


1. Start the app on your localhost
2. Install ngrok and run `ngrok http 3000`, this will allow your local host to accept incoming requests from slack API. copy the https url(ie: https://6b935d58.ngrok.io)
3. Send me your email and i will add you to slack app(cb_jarvis) and slack channel https://gauravchl.slack.com/ created for testing this bot.
4. Go to cb_jarvis's settings https://api.slack.com/apps/ACAB1TCTD/event-subscriptions and paste the copied ngrok url to Request URL field like: https://6b935d58.ngrok.io/webhook/slack , Click verify.

5. Go to https://gauravchl.slack.com/ and ping cb_jarvis. Try:

```
@cb_jarvis help 
@cb_jarvis list hangout 
@cb_jarvis create hangout, tomorrow at 5pm, Learn css 
Try different dates like: Friday at 2 pm
Note: It uses Chrono natural language date parser: https://github.com/wanasit/chrono 
```

### How to deploy to production 

1. Install app to codebuddies slack channel 
2. Set Request URL to https://codebuddies.org/webhook/slack
3. Copy oAuth token and set it to settings-production.json
4. Deploy it to production
5. Verify Request URL from slack app setting.
6. Test it!

**Comment by lpatmo** [comment]
https://www.youtube.com/watch?v=eDrZC0uv8hA&feature=youtu.be is super exciting. Thanks for recording the demo!

An idea/suggestion: After a hangout is successfully created, could the return message on Slack could contain the URL to the hangout that was just created? This way, the user could get to it to edit the details about the hangout, assuming they're logged into the site already. 

So the message might say:
`Hangout created successfully for 25 minutes, starting from Mon Sep. 3rd 2018. 5:00 pm Asia/Kolkota. Go to https://codebuddies.org/hangout/Tu84XrgG8kzTMxsNy to edit or join the hangout.`

I'm assuming that the payload from the site could include the post ID that's created, of course. If the URL isn't possible, we could say something like: 
`... Go to https://codebuddies.org/hangout` to find your hangout and edit it if necessary`

**Comment by gauravchl** [comment]
@lpatmo done 😎
screen shot 2018-09-04 at 6 22 17 pm

**Pull request merged**